### PR TITLE
add filesystem support

### PR DESCRIPTION
**Pull request by glerchundi**
* watch is not yet supported - see backends/fs/client.go for details
* integration should work
* exit if watch is enabled when fs is defined as backend

Tested with go 1.4

### CODE REVIEW

File: backends/fs/client.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	"strings"
+)
+
+type Client struct {
+	rootPath    string
+	maxFileSize int64
+}
+
+func NewFsClient() (*Client, error) {
+	// TODO: how to provide rootPath in order to pass tests?
```
Comment (bacongobbler) [inline backends/fs/client.go]: You can use `os.Setenv()` in the tests

File: backends/fs/client.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+func (c *Client) GetValues(keys []string) (map[string]string, error) {
+	vars := make(map[string]string)
+	for _, key := range keys {
+		keyPath := path.Join(c.rootPath, key)
+		if _, err := os.Stat(keyPath); os.IsNotExist(err) {
+			return vars, err
+		}
+
+		walk := func(fpath string, fi os.FileInfo, err error) error {
+			if (err != nil) {
```
Comment (bacongobbler) [inline backends/fs/client.go]: From https://golang.org/doc/effective_go.html#control-structures, parentheses in control structures are not allowed. You should've seen compile errors IIRC

File: backends/fs/client.go (hunk 1 above)
Comment (glerchundi) [inline backends/fs/client.go]: yeah, but os.Getenv(...) probably isn't the way to go. Maybe custom params for concrete backends in command line arguments? I put there just as a reminder, it's already being used in the `integration/test.sh` to get current testing folder.

Comment (bacongobbler) [inline backends/fs/client.go]: You could just pass in rootpath in the signature.

```
func NewFsClient(rootPath string, maxFileSize int64) (*Client, error) {
    ...
}
```

Comment (glerchundi) [inline backends/fs/client.go]: The idea was to provide a simple manner to use the existing integration tests and make them work with this new backend, nothing more. 'rootPath' could be achieved with prefix anyway.

**Comment by glerchundi** [comment]
Hi,

@bacongobbler @kelseyhightower 
Any update on this?

I'm going to explain why I found this useful. And I would like to know if there is a possibility to incorporate this on the main branch. If not, please tell me something!

I would like to keep secrets, certificates and everything related to security in etcd and then be able to use them inside a docker container which will be launched kubernetes:
- create a kubernetes volume, EmptyDir (from now on /volatile-vol/) one which is volatile by default (will be deleted when pod is destroyed).
- dump from a kvstore (etcd, zookeeper, whatever) to the `/volatile-vol`. For example, dump from `/secrets/databases/canary/db_password` to `/volatile-vol/DB_PASSWORD`
- use confd to create our configuration files based in `/volatile-vol/DB_PASSWORD`

Benefits:
- filter data from the backend and make available to the container only what matters to it.
- separation of concerns, I would like to hide to the container create how is structured our backend, just follow our patterns which define secrets would be available in `/volumes/secrets`
- there are already utilities which load environment variables from an envdir, for example, `s6-envdir`. We use it in this way: [`with-contenv`](https://github.com/just-containers/s6-overlay-builder/blob/v1.8.2/builder/overlay-rootfs/usr/bin/with-contenv). Dumping kvstore keys into directory facilitates patterns like this.

What do you think?

**Comment by bacongobbler** [comment]
Hmm, couple of thoughts here:

 - there should be a way for users to configure the max file size.
 - I know that there's interest in #183, so the maxFileSize and rootPath should be part of the backend's config for consistency, similar to https://github.com/kelseyhightower/confd/blob/master/backends/client.go#L34.
 - `--watch` should be [disabled entirely](https://github.com/kelseyhightower/confd/blob/master/config.go#L151-L154).

You'll also want to rebase off master so this PR can take advantage of the new test infra: https://github.com/kelseyhightower/confd/commit/1bc750778236c0bd4de55ab59bec600f754456b8

thanks!

**Comment by glerchundi** [comment]
Ok, I'll do it during this week,

Thanks!

**Comment by glerchundi** [comment]
Done.

- `fs` is included in the unsupported backends list.
- fs integration tests were included in the travis test infra.
- I really dislike the idea of mixing global opts with custom backend opts but, right now and without doing an important refactoring, I don't see how could be this done elegantly. So for now, I exposed `-fs-rootpath` and `-fs-maxfilesize`. A possible solution for the future could involve using `codegansta/cli` to create something like `confd [global opts] etcd/consul/fs/whatever [backend opts]`. If you guys @bacongobbler and @kelseyhightower think this is an interesting behaviour feature I could give it a try.

**Comment by bacongobbler** [comment]
> A possible solution for the future could involve using codegansta/cli to create something like confd [global opts] etcd/consul/fs/whatever [backend opts]. If you guys @bacongobbler and @kelseyhightower think this is an interesting behaviour feature I could give it a try.

I personally prefer to stick with the native flag package for handling option flags. That way we don't have to rewrite the entire CLI just to wrap around an opinionated lib. The standard library is pretty slick. :)

Regardless, I agree that there should be a better way to set up a backend's config options rather than through global flags. I'll keep thinking about how we can roll this out, but so far this implementation is about as clean as it gets without going too crazy.

**Comment by kelseyhightower** [comment]
@glerchundi Rebase please. I'm reviewing all the PRs today.

**Comment by glerchundi** [comment]
@kelseyhightower, done. travis is not detecting changes on PR, is it possible to force a new one?

**Comment by scalp42** [comment]
@glerchundi any chance you could rebase against master please?

@kelseyhightower any round of PR reviews coming soon please?

The use case for this backend is :+1:

**Comment by glerchundi** [comment]
@scalp42 done :-)

**Pull request closed without merging**